mne/annotations.py

```diff
@@ -647,6 +647,189 @@ def rename(self, mapping, verbose=None):
         return self
 
 
+class EpochAnnotationsMixin:
+    """Mixin class for Annotations in Epochs."""
+
+    @property
+    def annotations(self):  # noqa: D102
+        return self._annotations
+
+    @verbose
+    def set_annotations(self, annotations, on_missing='raise'):
+        """Setter for Epoch annotations from Raw.
+
+        This function simply copies over Raw annotations, and
+        does not handle offsetting the times based on first_samp
+        or measurement dates, since that is expected to occur in
+        Raw.set_annotations().
+
+        Parameters
+        ----------
+        annotations : instance of mne.Annotations | None
+            Annotations to set.
+        %(on_missing_ch_names)s
+
+        Returns
+        -------
+        self : instance of Epochs
+            The epochs object with annotations.
+
+        Notes
+        -----
+        Annotation onsets and offsets are stored as time in seconds (not as
+        sample numbers) so care must be taken when doing things like decimating
+        or resampling. If the user has an Epochs object that did not support
+        Annotations before, and the Epochs were resampled, then the Annotations
+        will not show up in the correct locations.
```

Comment:
I don't get this. As things are stored in seconds and not samples it becomes robust to resampling. I think this deserves some clarifications. I get that if raw sfreq is lost then it creates a problem but as is I find it confusing.

Reply:
How's this updated doc string instead?

Reply:
so by "an Epochs object that did not support annotations before" what we really mean is "an Epochs object saved before Annotation support was introduced" right?  (because it won't have an entry for `_raw_sfreq`)?  And by "will not show up in the correct locations" we mean "the Annotations onsets and durations will be shifted and stretched by the resampling/decimation ratio"?  Are those accurate rephrasings?

Reply:
Yep!

Reply:
ok, but... a quick test shows no problem, as @agramfort seemed to suspect:

```py
import mne

path = mne.datasets.kiloword.data_path() + '/kword_metadata-epo.fif'
epo = mne.read_epochs(path)

assert epo._decim == 1
epo.resample(epo.info['sfreq'] * 2)  # upsample, works with 0.5 too

ann = mne.Annotations(onset=[3,5], duration=[0.5, 0.6], description=['one', 'two'])
epo.set_annotations(ann)
for _ann in epo.annotations:
    print(_ann)
```

Can you provide an example @adam2392 where there *is* a problem with epochs saved prior to this PR?

Also along the way I noticed that `epo.set_annotations()` is part of the public API now, but its docstring says it "just copies annotations from Raw", so that docstring needs to be updated.

Reply:
Hmm.. okay will take a look. Perhaps along the way, we solved all the issues we initially foresaw?

Reply:
converting to Draft mode until we get this sorted out (with green CIs and 3 approvals I was worried someone might instinctively merge without looking too closely)

Reply:
```
path = mne.datasets.kiloword.data_path() + '/kword_metadata-epo.fif'
    epo = mne.read_epochs(path)
    assert epo._decim == 1
    ann = mne.Annotations(onset=[3,5], duration=[0.5, 0.6], description=['one', 'two'])
    epo.set_annotations(ann)
    epo.add_annotations_to_metadata()
    metadata = epo.metadata.copy()

    epo.resample(epo.info['sfreq'] * 1.5)  # upsample, works with 0.5 too
    epo.add_annotations_to_metadata(overwrite=True)
    new_metadata = epo.metadata  

    from pandas.testing import assert_frame_equal
    assert_frame_equal(metadata, new_metadata)
```

Okay this doesn't work because `events` are not altered when we do "resample". So therefore, `raw_sfreq` needs to be correctly set in order for this function to work.

Reply:
So this is one of the reasons why I had `set_annotations` as private now IIRC. It silently can introduce an error for downstream functions that leverage the Annotations. Currently, the only function we have is `add_annotations_to_metadata`. However, in the future we might have more.  

Now that I think about it though, I don't think it's necessarily an issue with `set_annotations` because that just puts Annotations into the `Epochs` class. It has to do with the fact that you can't really do anything useful w/ the Annotations if the Epochs object was resampled and you want to apply the original Annotations back to it. If you add to metadata, the Annotations placement are completely incorrect now.

To me this seems like an edge case that just needs to be documented.

Reply:
> Can you provide an example @adam2392 where there is a problem with epochs saved prior to this PR?

I think an Epochs constructed before this PR with `decim > 1`, saved, and loaded on the branch from this PR would show the problem, as `epochs.info['sfreq']` will not correspond to the sample rate of its `epochs.events`. I don't think anyone has tried this case in the above at least. I can try tomorrow unless someone beats me to it.

Reply:
> > Can you provide an example @adam2392 where there is a problem with epochs saved prior to this PR?
> 
> I think an Epochs constructed before this PR with `decim > 1`, saved, and loaded on the branch from this PR would show the problem, as `epochs.info['sfreq']` will not correspond to the sample rate of its `epochs.events`. I don't think anyone has tried this case in the above at least. I can try tomorrow unless someone beats me to it.

I added some additional clarification in the doc strings related to this issue. I can't find any test dataset that has the `decim > 1` property, so not sure how we could test this.

I also added a test case for `metadata` for Epochs resampled now with `_raw_sfreq` support that should match before/after resampling.

Reply:
> I can't find any test dataset that has the `decim > 1` property, so not sure how we could test this.

You could create one while on the `main` branch, save it, and add it to the testing data?

Reply:
> You could create one while on the main branch, save it, and add it to the testing data?

Indeed that's what I was suggesting for self-testing purposes, but we already have a no-data-added way to do it with `monkeypatch` in the tests. I'll push something shortly that I think should help, namely, 1) expose the problem, 2) add a warning when it might be a problem. (2) is enabled because it should only possibly be an issue when working with `EpochsFIF`, so when people `set_annotations` on those we can warn them that they should regenerate them if they originally used decimation or resampling. This alongside adding a `verbose` kwarg (so that people can `verbose='error'` if they know it's not an issue) I think makes us safe enough.

Reply:
Okay @adam2392 @drammock @agramfort feel free to look at 05cbbbd -- the test should make clear both the corner case and the new behavior I added to warn people about it.

Reply:
Ah thanks @larsoner